[#10] Port patterns from config objects to config entities

juampynr opened the pull request:
This is a work in progress for https://github.com/md-systems/pathauto/issues/110. It converts patterns from config objects to config entities.

Given the following setup:

![image](https://cloud.githubusercontent.com/assets/108130/12035192/6ea37f8a-ae3a-11e5-92ec-43a03b2af9c9.png)

After running database updates, we get this message:

![selection_003](https://cloud.githubusercontent.com/assets/108130/12035207/7f5327c2-ae3a-11e5-801d-767c9462a278.png)

Finally, when we open the patterns section, we can see that they have been created:

![selection_004](https://cloud.githubusercontent.com/assets/108130/12035211/90b6fa52-ae3a-11e5-82d1-5c86e3492a25.png)
## TODO
- [x] Process forum entities.
- [x] Figure out how patterns for specific languages are created at 8.x-1.x and then adjust the database update so it creates the respective config entities.

juampynr committed:
[#10] Port patterns from config objects to config entities

Berdir commented:
You can see the structure for languages in the schema. it's another languages key below the bundle, next to default there, which then has the per-language patterns.

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,101 @@ function pathauto_update_8001() {
+  $delete_config = TRUE;
+
+  if (!\Drupal::service('module_handler')->moduleExists('ctools')) {
+    throw new UpdateException('Please, install Chaos tools suite (https://www.drupal.org/project/ctools) before running this databsae update.');
```
I think there's a rule to not use Please in user-facing texts :)

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,101 @@ function pathauto_update_8001() {
+  foreach ($patterns as $entity_type => $pattern_config) {
+    if (!array_key_exists($entity_type, $entity_info)) {
+      // We found an entity type which we don't know how to process. Report it.
+      $messages[] = t('Entity of type @type was not processed.', array('@type' => $entity_type));
```
I'm not sure what to do about forum, to be honest. I'm OK with leaving that out, as far as I know, nobody actually tried to use it in 8.x, it has no tests, so it is very likely broken anyway.

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,101 @@ function pathauto_update_8001() {
+      $delete_config = FALSE;
+      continue;
+    }
+    $entity_label = (string) $entity_info[$entity_type]->get('label');
```
getLabel()

juampynr committed:
[#10] Add support to port language-dependant patterns

juampynr commented:
@Berdir: I have added support for language-dependent patterns plus fixes for your previous peer review. This is how it looks like now:

Given the following config at `pathauto.patterns`:

```
array (
  'patterns' => 
  array (
    'node' => 
    array (
      'default' => '/content/[node:title]',
      'bundles' => 
      array (
        'article' => 
        array (
          'default' => 'articles/[node:title]',
        ),
        'article_en' => 
        array (
          'default' => '',
        ),
        'article_fr' => 
        array (
          'default' => 'lesarticles/[node:title]',
        ),
        'article_es' => 
        array (
          'default' => 'articulos/[node:title]',
        ),
        'page' => 
        array (
          'default' => 'pages/[node:title]',
        ),
        'page_en' => 
        array (
          'default' => '',
        ),
        'page_fr' => 
        array (
          'default' => '',
        ),
        'page_es' => 
        array (
          'default' => '',
        ),
      ),
    ),
    'taxonomy_term' => 
    array (
      'default' => '/[term:vocabulary]/[term:name]',
      'bundles' => 
      array (
        'bo' => 
        array (
          'default' => '',
        ),
        'leswitidr' => 
        array (
          'default' => '',
        ),
        'leswitidr_en' => 
        array (
          'default' => '',
        ),
        'leswitidr_fr' => 
        array (
          'default' => '',
        ),
        'leswitidr_es' => 
        array (
          'default' => 'tugs/[term:name]',
        ),
        'tags' => 
        array (
          'default' => 'togs/[term:name]',
        ),
      ),
    ),
    'forum' => 
    array (
      'default' => '/[term:vocabulary]/[term:name]',
    ),
    'user' => 
    array (
      'default' => '/users/[user:name]',
    ),
  ),
  '_core' => 
  array (
    'default_config_hash' => 'siBCX8BOF-Gu6C0RlHWM8LUecYXeSYNM9o9FgPGcPrA',
  ),
)
```

After running the database update, we get the following patterns in the new architecture:

![selection_001](https://cloud.githubusercontent.com/assets/108130/12050695/493788d0-aef9-11e5-987e-b8e19f6b2737.png)

What do you think? I tested this by following this steps:
1. Install pathauto 8.x.1-x. Create patterns for different languages.
2. Replace 8.x-1.x by 110-update-path-patterns. Install Ctools.
3. Run database updates.

Berdir commented:
Interesting. I know that there is a bug report that language patterns don't work. And as reported there, it's because the UI creates them incorrectly, not how the schema and code excepts it.

So the question is, should we migrate the possibly  existing but not working config or the config that's supposed to exist :)

Either way, I don't really care that much, I don't have an multlingual sites myself that use pathauto and I doubt it fully works anyway. I'll review this one more time and then I'll merge it.

juampynr commented:
> Interesting. I know that there is a bug report that language patterns don't work. And as reported there, it's because the UI creates them incorrectly, not how the schema and code excepts it.

Do you mean [this one](https://github.com/md-systems/pathauto/issues/92)?

> So the question is, should we migrate the possibly existing but not working config or the config that's supposed to exist :)

Yeah, let's migrate them and fix the UI. I will now test and submit patches for https://github.com/md-systems/pathauto/issues/92.

> Either way, I don't really care that much, I don't have an multlingual sites myself that use pathauto and I doubt it fully works anyway. I'll review this one more time and then I'll merge it.

Cool, I will start testing and debugging multilingual patterns.

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,137 @@ function pathauto_update_8001() {
+  $entity_manager = \Drupal::service('entity.manager');
+  $entity_type_manager = \Drupal::service('entity_type.manager');
+  $language_manager = \Drupal::service('language_manager');
+  $entity_info = $entity_manager->getDefinitions();
```
$entity_info is a Drupal 7 thing. Lets use $entity_types.

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,137 @@ function pathauto_update_8001() {
+
+    if (isset($entity_patterns['default'])) {
+      // This is a pattern for an entity type, such as "node".
+      $entity = entity_create('pathauto_pattern', [
```
entity_create is deprecated, use PathautoPattern::create()

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,137 @@ function pathauto_update_8001() {
+          // an unknown bundle or langcode. Let's figure it out.
+          $matches = NULL;
+          $langcode = NULL;
+          preg_match('/^(.*)_([a-z-]*)$/', $bundle, $matches);
```
Does this work with a bundle like some_thing? I guess so.. Instead of *, maybe use + so that you require at least one character.

juampynr replied:
Yes it does, I tested it manually. See the following:

```
php > $a = 'foo_bar_de';
php > unset($matches);var_dump(preg_match('/^(.*)_([a-z-]*)$/', $a, $matches));var_dump($matches);
int(1)
array(3) {
  [0] =>
  string(10) "foo_bar_de"
  [1] =>
  string(7) "foo_bar"
  [2] =>
  string(2) "de"
}
```

Berdir commented on pathauto.install:
```diff
@@ -51,3 +52,137 @@ function pathauto_update_8001() {
+  }
+
+  // 3. Delete the old configuration object that stores patterns only if all
+  $config->delete();
```
comment is wrong, it's not conditional anymore.

juampynr commented on pathauto.install:
```diff
@@ -51,3 +52,137 @@ function pathauto_update_8001() {
+            'langcodes' => [ $langcode => $langcode ],
+            'negate' => FALSE,
+            'context_mapping' => [
+              'language' => $entity_type . ':' . $entity_type_manager->getDefinition($entity_type)->getKey('langcode') . ':language',
```
@Berdir isn't the above line doing the same as the following?

```
          $new_definition = new ContextDefinition('language', 'Language');
          $new_context = new Context($new_definition);
          $pattern->addContext($language_context, $new_context);
```

I will test creating a pattern through the interface anyway to see if there is a difference in the resulting config entity.

juampynr committed:
[#10] Add language context to patterns

Plus:
* Small improvements suggested by Berdir.
* Make sure that the langcode is a valid machine-name. For example,
  Chinese traditional has a langcode zh-hant, which is not valid for a
  machine name.

juampynr commented:
@Berdir I have implemented all your feedback, plus verified that now each pattern has the ctools context when they have a language condition.

Berdir commented:
Nice, merged!
